src/glyph.ts

```diff
@@ -371,10 +371,10 @@ export class Glyph extends Element {
       x_max: this.metrics.x_max * this.scale * this.metrics.scale,
       width: this.bbox.getW(),
       height: this.bbox.getH(),
-      scale: 1,
-      x_shift: 0,
-      y_shift: 0,
-      outline: [],
+      scale: this.metrics.scale,
+      x_shift: this.metrics.x_shift,
+      y_shift: this.metrics.y_shift,
+      outline: this.metrics.outline,
```

Comment:
should this be a copy of the Array, so it doesn't mutate?

Reply:
But the attributes above are not one to one copies of metric:
```js
return {
      x_min: this.metrics.x_min * this.scale * this.metrics.scale,
      x_max: this.metrics.x_max * this.scale * this.metrics.scale,
      width: this.bbox.getW(),
      height: this.bbox.getH(),
      scale: this.metrics.scale,
      x_shift: this.metrics.x_shift,
      y_shift: this.metrics.y_shift,
      outline: this.metrics.outline,
    };
```
I did not went into this investigation. I find specially strange the two scales: `this.scale` and `this.metrics.scale`
@mscuthbert do you know why is that so?

Reply:
@0xfe can you answer the question?

Reply:
`scale` is a default scaling factor for the class, and used by the rendering context, and `this.metrics.scale` is a font-specific override loaded from the font metrics files. E.g., this allows two `flat` symbols in different fonts with different sizes to scale consistently within the context.

Reply:
For `scale` it may be better to return `this.metrics.scale * this.scale` -- either way, how does this affect the visual tests? Unclear to me what the downstream effects of this change are.

Reply:
No, the number of change go beyond the threshold

Reply:
// [VexFlow](http://vexflow.com) - Copyright (c) Mohit Muthanna 2010.
 //
 // ## Description
 // Implements time signatures glyphs for staffs
 // See tables.js for the internal time signatures
 // representation
 
 import { Vex } from './vex';
 import { Glyph } from './glyph';
 import { StaveModifier } from './stavemodifier';
+import { TimeSignatureGlyph } from './timesigglyph';
 
-const assertIsValidFraction = (timeSpec) => {
+export interface TimeSignatureInfo {
+  glyph: Glyph;
+  line?: number;
+  num: boolean;
+}
+
+const assertIsValidFraction = (timeSpec: string) => {
   const numbers = timeSpec.split('/').filter((number) => number !== '');
 
   if (numbers.length !== 2) {
     throw new Vex.RERR(
       'BadTimeSignature',
       `Invalid time spec: ${timeSpec}. Must be in the form "<numerator>/<denominator>"`
     );
   }
 
   numbers.forEach((number) => {
     if (isNaN(Number(number))) {
       throw new Vex.RERR('BadTimeSignature', `Invalid time spec: ${timeSpec}. Must contain two valid numbers.`);
     }
   });
 };
 
 export class TimeSignature extends StaveModifier {
-  static get CATEGORY() {
+  point?: number;
+  bottomLine?: number;
+  protected timeSig?: TimeSignatureInfo;
+
+  protected validate_args: boolean;
+
+  topLine?: number;
+
+  static get CATEGORY(): string {
     return 'timesignatures';
   }
 
-  static get glyphs() {
+  static get glyphs(): Record<string, { code: string; point: number; line: number }> {
     return {
       C: {
         code: 'timeSigCommon',
         point: 40,
         line: 2,
       },
       'C|': {
         code: 'timeSigCutCommon',
         point: 40,
         line: 2,
       },
     };
   }
 
-  constructor(timeSpec = null, customPadding = 15, validate_args = true) {
+  constructor(timeSpec?: string, customPadding = 15, validate_args = true) {
     super();
     this.setAttribute('type', 'TimeSignature');
     this.validate_args = validate_args;
 
-    if (timeSpec === null) return;
-
+    if (!timeSpec) return;
     const padding = customPadding;
 
     this.point = this.musicFont.lookupMetric('digits.point');
     const fontLineShift = this.musicFont.lookupMetric('digits.shiftLine', 0);
     this.topLine = 2 + fontLineShift;
     this.bottomLine = 4 + fontLineShift;
     this.setPosition(StaveModifier.Position.BEGIN);
-    this.setTimeSig(timeSpec);
-    this.setWidth(this.timeSig.glyph.getMetrics().width);
+    this.timeSig = this.parseTimeSpec(timeSpec);
+    this.setWidth(this.getTimeSig().glyph.getMetrics().width!);
     this.setPadding(padding);
   }
 
-  getCategory() {
+  getCategory(): string {
     return TimeSignature.CATEGORY;
   }
 
-  parseTimeSpec(timeSpec) {
+  parseTimeSpec(timeSpec: string): TimeSignatureInfo {
     if (timeSpec === 'C' || timeSpec === 'C|') {
       const { line, code, point } = TimeSignature.glyphs[timeSpec];
       return {
         line,
         num: false,
         glyph: new Glyph(code, point),
       };
     }
 
     if (this.validate_args) {
       assertIsValidFraction(timeSpec);
     }
 
     const [topDigits, botDigits] = timeSpec.split('/').map((number) => number.split(''));
 
     return {
       num: true,
       glyph: this.makeTimeSignatureGlyph(topDigits, botDigits),
     };
   }
 
-  makeTimeSignatureGlyph(topDigits, botDigits) {
-    const glyph = new Glyph('timeSig0', this.point);
-    glyph.topGlyphs = [];
-    glyph.botGlyphs = [];
-
-    let topWidth = 0;
-    for (let i = 0; i < topDigits.length; ++i) {
-      const num = topDigits[i];
-      const topGlyph = new Glyph('timeSig' + num, this.point);
-
-      glyph.topGlyphs.push(topGlyph);
-      topWidth += topGlyph.getMetrics().width;
-    }
-
-    let botWidth = 0;
-    for (let i = 0; i < botDigits.length; ++i) {
-      const num = botDigits[i];
-      const botGlyph = new Glyph('timeSig' + num, this.point);
-
-      glyph.botGlyphs.push(botGlyph);
-      botWidth += botGlyph.getMetrics().width;
-    }
-
-    const width = topWidth > botWidth ? topWidth : botWidth;
-    const xMin = glyph.getMetrics().x_min;
-
-    glyph.getMetrics = () => ({
-      x_min: xMin,
-      x_max: xMin + width,
-      width,
-    });
-
-    const topStartX = (width - topWidth) / 2.0;
-    const botStartX = (width - botWidth) / 2.0;
-
-    const that = this;
-    glyph.renderToStave = function renderToStave(x) {
-      let start_x = x + topStartX;
-      for (let i = 0; i < this.topGlyphs.length; ++i) {
-        const glyph = this.topGlyphs[i];
-        Glyph.renderOutline(
-          this.context,
-          glyph.metrics.outline,
-          glyph.scale,
-          start_x + glyph.x_shift,
-          this.stave.getYForLine(that.topLine)
-        );
-        start_x += glyph.getMetrics().width;
-      }
-
-      start_x = x + botStartX;
-      for (let i = 0; i < this.botGlyphs.length; ++i) {
-        const glyph = this.botGlyphs[i];
-        that.placeGlyphOnLine(glyph, this.stave, glyph.line);
-        Glyph.renderOutline(
-          this.context,
-          glyph.metrics.outline,
-          glyph.scale,
-          start_x + glyph.x_shift,
-          this.stave.getYForLine(that.bottomLine)
-        );
-        start_x += glyph.getMetrics().width;
-      }
-    };
-
+  makeTimeSignatureGlyph(topDigits: string[], botDigits: string[]): Glyph {
+    const glyph = new TimeSignatureGlyph(this, topDigits, botDigits, 'timeSig0', this.point!);
     return glyph;
   }
 
-  getTimeSig() {
-    return this.timeSig;
+  getTimeSig(): TimeSignatureInfo {
+    return this.timeSig!;
   }
 
-  setTimeSig(timeSpec) {
+  setTimeSig(timeSpec: string): this {
     this.timeSig = this.parseTimeSpec(timeSpec);
     return this;
   }
 
-  draw() {
+  draw(): void {
     if (!this.x) {
       throw new Vex.RERR('TimeSignatureError', "Can't draw time signature without x.");
     }
 
     if (!this.stave) {
       throw new Vex.RERR('TimeSignatureError', "Can't draw time signature without stave.");
     }
 
     this.setRendered();
-    this.timeSig.glyph.setStave(this.stave);
-    this.timeSig.glyph.setContext(this.stave.context);
-    this.placeGlyphOnLine(this.timeSig.glyph, this.stave, this.timeSig.line);
-    this.timeSig.glyph.renderToStave(this.x);
+    this.timeSig!.glyph.setStave(this.stave);
+    this.timeSig!.glyph.setContext(this.stave.getContext());
+    this.placeGlyphOnLine(this.timeSig!.glyph, this.stave, 2);
+    this.timeSig!.glyph.renderToStave(this.x);
   }
 }

Reply:
@0xfe there you are, generated with diff.

Reply:
> For `scale` it may be better to return `this.metrics.scale * this.scale` -- either way, how does this affect the visual tests? Unclear to me what the downstream effects of this change are.

Done, no change on images.

Reply:
Thanks for the diff!